### ISSUE TITLE
net/http: use custom backend IP address in https request with proxy

### ISSUE THREAD

**Issue by liuxingbaoyu** [golang/go#45460]
### What version of Go are you using (`go version`)?

<pre>
go version go1.16 windows/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
set GO111MODULE=auto
set GOARCH=amd64
set GOBIN=
set GOCACHE=C:\Users\流星暴雨\AppData\Local\go-build
set GOENV=C:\Users\流星暴雨\AppData\Roaming\go\env
set GOEXE=.exe
set GOFLAGS=
set GOHOSTARCH=amd64
set GOHOSTOS=windows
set GOINSECURE=
set GOMODCACHE=E:\Eternal\go\pkg\mod
set GONOPROXY=
set GONOSUMDB=
set GOOS=windows
set GOPATH=E:\Eternal\go;
set GOPRIVATE=
set GOPROXY=https://goproxy.io,direct
set GOROOT=C:\Program Files\Go
set GOSUMDB=sum.golang.org
set GOTMPDIR=
set GOTOOLDIR=C:\Program Files\Go\pkg\tool\windows_amd64
set GOVCS=
set GOVERSION=go1.16
set GCCGO=gccgo
set AR=ar
set CC=gcc
set CXX=g++
set CGO_ENABLED=1
set GOMOD=
set CGO_CFLAGS=-g -O2
set CGO_CPPFLAGS=
set CGO_CXXFLAGS=-g -O2
set CGO_FFLAGS=-g -O2
set CGO_LDFLAGS=-g -O2
set PKG_CONFIG=pkg-config
set GOGCCFLAGS=-m64 -mthreads -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=C:\Users\ 流星暴雨\AppData\Local\Temp\go-build1194816388=/tmp/go-build -gno-record-gcc-switches
</pre></details>

### What did you do?

I want to use a proxy to access HTTPS websites and customize the backend IP address

### What did you expect to see?

Is there some way to customize the URL of the CONNECT request

### What did you see instead?
Both CONNECT request and SSL handshake use `cm.targetAddr`

https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/net/http/transport.go#L1667-L1672
https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/net/http/transport.go#L1725-L1729
https://github.com/golang/go/blob/2ebe77a2fda1ee9ff6fd9a3e08933ad1ebaea039/src/net/http/transport.go#L1838-L1846

**Comment by seankhliao** [comment]
Unlike many projects, the Go project does not use GitHub Issues for general discussion or asking questions. GitHub Issues are used for tracking bugs and proposals only.

For questions please refer to https://github.com/golang/go/wiki/Questions

**Comment by liuxingbaoyu** [comment]
@seankhliao Oh no, I'm sorry that I didn't express it clearly. I read the relevant code and did not find a way to customize the CONNECT request. Perhaps this requires a new API.

**Comment by seankhliao** [comment]
what exactly do you want to do? change the address the of the proxy while keeping the hostname?

**Comment by liuxingbaoyu** [comment]
When an HTTPS request is sent through an HTTP proxy, it will first send a CONNECT request to establish a TCP connection, and then it will start a handshake with the SSL of the website. If a website has multiple resolution addresses, such as using a CDN, you can use the specified backend IP to access the website by specifying the CONNECT URL. I hope there is an API that can customize the CONNECT URL or customize the entire CONNECT request.

**Comment by seankhliao** [comment]
Right, different values for `CONNECT xxx` (for connecting) and `Host: yyy` (for TLS)

**Comment by seankhliao** [comment]
cc @bradfitz @neild @fraenkel

**Comment by rsc** [comment]
The code that sets up the CONNECT does:

	connectReq := &Request{
		Method: "CONNECT",
		URL:    &url.URL{Opaque: cm.targetAddr},
		Host:   cm.targetAddr,
		Header: hdr,
	}

If hdr["Host"] is set, then it sounds like we should use it instead of cm.targetAddr to set up connectReq.Host. That is, add:

    if host, ok := hdr["Host"]; ok {
        if len(host) >= 1 {
            connectReq.Host = host[0]
        } else {
            connectReq.Host = ""
        }
    }

Does anyone have any objections to this?

**Comment by rsc** [comment]
This proposal has been added to the [active column](https://golang.org/s/proposal-status#active) of the proposals project
and will now be reviewed at the weekly proposal review meetings.
— rsc for the proposal review group

**Comment by fraenkel** [comment]
Sorry but I am confused on what we are trying to solve.

We have a request to some target site, www.target.org, and we are going through some Proxy. The actual request has a URL/host of www.target.org. It really doesn't matter what the host header is. The Request will ignore the Host header and generate its own based, https://github.com/golang/go/blob/master/src/net/http/request.go#L570

A proxy is required to ignore the host header if it receives an absolute uri, https://tools.ietf.org/html/rfc7230#section-5.4
```
When a proxy receives a request with an absolute-form of
   request-target, the proxy MUST ignore the received Host header field
   (if any) and instead replace it with the host information of the
   request-target.
```

**Comment by rsc** [comment]
@fraenkel, that's for a standard HTTP proxy, not for some custom thing serving CONNECT requests.
Do you object to providing this functionality? It seems unobjectionable to us.

**Comment by fraenkel** [comment]
@rsc sorry my bad.
I missed the fact that we were talking about the ProxyConnectHeaders.
You will also have to reset `connectReq.URL.Opaque = ""` to have it appear on the CONNECT line.

**Comment by fraenkel** [comment]
Probably easier to rewrite the url since the host is only used of there is no opaque bit in the url.

**Comment by fraenkel** [comment]
Sorry I confused myself again.  Ignore what I said. I need to double check my test case which didn't see the host header on the other side.

**Comment by fraenkel** [comment]
I forgot that ReadRequest deletes the host header. So as long as the other side isn't using Go, this should work.

**Comment by rsc** [comment]
Based on the discussion above, this proposal seems like a **[likely accept](https://golang.org/s/proposal-status#likely-accept)**.
— rsc for the proposal review group

**Comment by rsc** [comment]
No change in consensus, so **[accepted](https://golang.org/s/proposal-status#accepted)**. 🎉
This issue now tracks the work of implementing the proposal.
— rsc for the proposal review group

**Comment by neild** [comment]
```
if host, ok := hdr["Host"]; ok {
    if len(host) >= 1 {
        connectReq.Host = host[0]
    } else {
        connectReq.Host = ""
    }
}
```

I don't believe this is correct.

[RFC 9110 9.3.6](https://www.rfc-editor.org/rfc/rfc9110#name-connect) states that a proxy should connect to the "server identified by the request target". It doesn't say anything about the `Host` header that I can see, and I can't find any examples of using different values in the target and the `Host` header.

I did find a [2009 ietf-http-wg thread](https://lists.w3.org/Archives/Public/ietf-http-wg/2009AprJun/0696.html) discussing the role of the `Host` header in `CONNECT` requests which seems to have concluded that the header is required, but that proxies should either ignore it or validate that it exactly matches the request target.

That said, the general idea of setting the target of the proxy request via the proxy connect headers seems reasonable; we should just set both the request target and the host header.
